### ISSUE TITLE
Custom schema and its parsing

### ISSUE THREAD

**Issue by AistisT** [asyncapi/parser-js#33]
This might be daft, but I'm a bit puzzled.
How does one uses custom schema that is based on AsyncApi one. I assume I need a custom parser, but what's the role of the parser?

After registering custom parser, how does one call parse? Parser.parse() has no clue about custom parsers there were registered.
Is it expected for iterateDocument() to be used instead of parse with custom parser?

Also I'm bit uncertain about how exactly one points to a schema when registering it.
parser.registerSchemaParser(
        [
          "application/vnd.aai.asyncapi;version=2.0.0",
          "application/vnd.aai.asyncapi+json;version=2.0.0",
          "application/vnd.aai.asyncapi+yaml;version=2.0.0",
          "application/schema;version=draft-07",
          "application/schema+json;version=draft-07",
          "application/schema+yaml;version=draft-07"
        ],

this is clearly not a path to a file holding a schema. I've looked through all schemas, and apart from version, they do not contain any vnd.aa.xxx fields, thus I am clueless on how it matches to a file schema.

Would really appreciate a little of guidance on how to register a custom schema (from a local file path) and its parser and how parse documents with it.

**Comment by fmvilas** [comment]
Oh sorry for the confusion. This is not yet documented but it works more or less like this:

```js
const parser = require('asyncapi-parser');
parser.registerSchemaParser([
  'your-mime-type'
], yourFunctionToParseYourSchemaFormat);

function yourFunctionToParseYourSchemaFormat ({ message, defaultSchemaFormat }) {
  // Transform your schema format to JSON Schema
}
```

This is an example of how our OpenAPI parser works: https://github.com/asyncapi/parser-js/blob/master/lib/schema-parsers/openapi.js.

That said, if you want to extract your parser to another file, move the `yourFunctionToParseYourSchemaFormat` to another file and require it. Like this:


```js
const parser = require('asyncapi-parser');
const yourFunctionToParseYourSchemaFormat = require('path/to/yourFunctionToParseYourSchemaFormat');

parser.registerSchemaParser([
  'your-mime-type'
], yourFunctionToParseYourSchemaFormat);
```

Then use `parser` to parse and validate your messages. Does it help?

**Comment by AistisT** [comment]
It does help somewhat, so schemas are registered by mime type? But how does it know where to look for files matching it.
Say we have following project directory  setup:
Root containing src folder which contains templates and schema folders.
I am able to pass template folder when creating generator, but parser is not part of generator nor the paths are the same. So how would it know to look in root/src/template instead of root/src or root, which ever is default path at a time.

You mentioned use parser to parse and validate, but exactly how. What api's am I expected to call, because from what I've seen i cannot use `parser.parse()` as it will first of check of asyncapi schema versions which is not what I want. It also does not take into consideration of custom parser.

**Comment by AistisT** [comment]
I've been playing around, and I'm slowly getting a grasp. Still unsure how schemas are picked up based on mime types when registering them. 
I now know what parser should do, transform to asyncapi expected schema, which I think is wrong, that should not be forced upon. Its extra work for anyone that wants to change asyncapi field to myschema field, let json validator do its validating.

Why enforce this branding that noone can see that creates additional work?

**Comment by fmvilas** [comment]
Sorry for the late reply. The reason we're transforming to JSON Schema is that this will become easier for the rest of the world to develop tooling because, independently of the original format, tooling will only have to support JSON Schema. So no, it's not branding, we're enforcing you to do the job just so the rest of the world don't have to do it, which I think is fair, right?

Also, if you do that, you'll benefit from existing tooling like documentation and code generators. We're just putting the responsibility in the right place.

**Comment by AistisT** [comment]
Yes there is nothing wrong with transforming to JSON, from what I recall it's done by third party library and it doesn't care what json it is as long its valid json, and then it goes though json schema validation.

The problem is before it gets to transformation and schema validation. If our yaml doesn't contain asyncapi field or specific versions, it won't get to that part. That's what I've meant that people are forced to take their yaml, transform to asincapi version of yaml schema, then it eventually gets transformed to json at which point it doesn't care about any specific schema.

That should not be the case. I realise project is still early in its stages and documentation is bit scarce, but we found it easier to branch from asincapi instead of using it and providing our own parser, registering it and figuring how to use it, above example sadly didn't provide much help as I still had no clue how to actually indicate schema files for custom parser to use. Parser should not be necessary. 

Allow people provide their own yaml schema and versions they support, the second you take out custom validation of asincapi fields and its versions, any schema can be used with the library, which is most likely based on asincapi.

Or please export private methods that deal with validation and parsing of the yaml file, to allow people override default parse method that validates asincapi field which most people will take out and replace with their own branding/versioning.

In short, any valid yaml that can be parsed to json, should be allowed, allow generic ways for people to provide versions numbers, don't validate/read on specific yaml fields, make it fully generic.

**Comment by fmvilas** [comment]
> Yes there is nothing wrong with transforming to JSON

I'm not saying it's transforming to JSON. I'm saying message payloads are transformed to JSON Schema, which is different than JSON.

> If our yaml doesn't contain asyncapi field or specific versions, it won't get to that part.

Of course, because it would mean it's not a valid AsyncAPI document. Are you trying to parse any kind of YAML/JSON file?

> then it eventually gets transformed to json at which point it doesn't care about any specific schema

That's not true. After the file is transformed to JSON, it's then validated applying our JSON Schema [schema](https://github.com/asyncapi/asyncapi-node/tree/master/schemas).

--

I suspect that we're not talking about the same kind of schemas. Are you looking to validate AsyncAPI 1.x documents with this parser? If that's the case, we're not implementing support for it because it's a deprecated version already and this parser started after v2.0.0 was launched. If your AsyncAPI documents are using v1.x you can migrate them using the automated converter, either in [Javascript](https://github.com/asyncapi/converter) or [Go](https://github.com/asyncapi/converter-go).

PS: It's becoming really hard to understand you. Could you please elaborate a bit more? Otherwise, it's really hard to help.

**Comment by AistisT** [comment]
My apologies for my  incoherent rambling.

We took asincapi 2.0.0 schema and modified it to suit our needs, as no schema can be suitable for everyone.  It's still 80% of asincapi schema, but its not exact schema any more , which is the problem I am trying to point out.

Any big and significant project, will take asincapi schema and modify it and that should be ok and encouraged.

> That's not true. After the file is transformed to JSON, it's then validated applying our JSON Schema schema.

Any schema can be provided there, in our case we provided our based on asincapi one. And it was validated without any issues.
Which is the point I am trying to make, let people provide their own Schemas, in similar fashion, you register asincapi schemas, which is then used for validation and generation. It should be ok for that schema not to use asincapi or any other field, instead have a generic way for people to do versioning, and branding of their schemas.

For example one of the options passsed into parse method, could be equivalent to asincapi field that is used to obtain version and perform any basic validation to check that it is expected file.

**Comment by fmvilas** [comment]
There are 2 things here:

1. `parser.registerSchemaParser` doesn't do what you think it does. This function is to register "schema" parsers. Examples of schemas are JSON Schema, Avro, RAML (support is coming soon), Protobuf (not supported yet), and XSD (not supported yet). These are ways to define your **message payloads, not the whole AsyncAPI file.** It can be confusing because both are actually schemas.
1. You should never create your custom schema from the AsyncAPI schema. Otherwise, what's the point of following a standard then? If you need to extend AsyncAPI to suit your needs then go for [specification extensions](https://github.com/asyncapi/asyncapi/blob/master/versions/2.0.0/asyncapi.md#specificationExtensions). You can use extensions almost everywhere in your AsyncAPI document and they are custom objects with free-form. **They will not be validated**. See the following example:

```yaml
asyncapi: 2.0.0
...
x-a-feature-we-need:
  something: 5
  anotherThing:
    test: true
```

This is completely valid AsyncAPI because the extension is starting with `x-` and though it will not be validated. We plan on adding support for extension validation in the future, so you could eventually do something like `parser.registerExtension('x-a-feature-we-need', fnValidate)`. In the meantime, you can validate extensions yourself in your code **after** parsing/validating the AsyncAPI document with this parser.

Does that make sense?

**Comment by AistisT** [comment]
It does make sense, but isn't validation done by avj and its generic json schema validation. So if we define the fields we want in json schema, it will be validated by avj json schema validation, which seems superior to non validated extension fields. I might be mistaken here, but most of parsing code is generic.

Following standards is important, but we are defining our own standard for the industry, thus the need to modify the json schema. We achieved everything we wanted, but that resulted in modifying source code and branching from original source, which might or might not be a good thing.

I just feel that there should be a little more freedom in using asyncapi libraries as its really good at what it does.

**Comment by jonaslagoni** [comment]
I can understand your point of view @AistisT however as @fmvilas stated, AsyncAPI provide a number of ways to specify your own modifications to the AsyncAPI document, while still complying with the standard. And when the functionality of adding custom extension parsers are added you will be able to do what I am assuming you are doing now in the official parser. 

It all comes down to the complexity of the parser and the maintainability and in my mind it makes little sense for this parser to be able to parse anything other then AsyncAPI schemas, given there are the option of extensions and upcoming custom parsers for those extensions.

**Comment by AistisT** [comment]
Apart from adding UML parsing/generating, we really haven't done much of custom processing.

As I've said we had to do very little for our modified schema to be parsed and validated.

Yaml is converted by standard JSON library and converted json is validated by ajv.
`   return JSON.parse(xxxYAMLorJSON);`
The source code is generic and is not tied to asyncapi schema.
`const validate = ajv.compile(xxxapi[js.xxx]);
    const valid = validate(js);`
 Apart from the start of parse function where it checks for asincapi version. 
`  if (!js.xxx || !xxxapi[js.xxx]) {
      throw new ParserErrorUnsupportedVersion(
        `xfs4iot version is missing or unsupported: ${js.xfs4iotapi}.`,
        js
      );
    }
`

I am not saying that any more complexity should be added, I think some of it should be removed to allow more freedom for developers, that cannot follow or use asincapi schema.

I understand if that's is not the object of asincapi project.
And I realise we gone bit off topic from my original question.

**Comment by jonaslagoni** [comment]
> I am not saying that any more complexity should be added, I think some of it should be removed to allow more freedom for developers, that cannot follow or use asincapi schema.

I think from AsyncAPI's point of view this use-case would never exist, since the AsyncAPI parser should not be used for parsing anything other then official AsyncAPI schemas. As you pointed out `Following standards is important, but we are defining our own standard for the industry, thus the need to modify the json schema` when you are defining your own standard why would you want to depend on changes from AsyncAPI? You said 80% is the AsyncAPI standard thus shouldn't you just use AsyncAPI as the standard instead and add what ever 20% as extensions to your use-case, instead of trying to come up with a new standard just for your use-case? :slightly_smiling_face: 
This would also enable you to use any of the other tools i.e. the generator :smile:

**Comment by AistisT** [comment]
Well this is very very start of this project and very beginning of schema standard, 80% is just a random number I made up, I do believe its around that ATM, but it might change in the future.

We are using generator, obviously we are changing templates, but generator allows us to  specify custom templates. We really didn't have to modify the generator source code to work with our derived schema. 

I didn't know about extensions to yaml until just now, and since no one else mentioned it within the team, I assume no one else knew either. 

I still believe that modifying schema while following json standards, and allowing json to validate and parse is superior approach to extension fields which are excluded from validation or have to have its own validators.

I can see why this might not be use case for AsincAPI, though this is first time I am using it and thus I assumed that using it with your own schema is valid use case.
